python/ops.py

```diff
@@ -13,12 +13,37 @@
 # limitations under the License.
 """FFCC TensorFlow ops."""
 import math
-import sys
-
+import itertools
 import numpy as np
 import tensorflow.compat.v1 as tf
 
 
+
+EPS = tf.constant(1e-9, dtype=tf.float32)
+
+
+def edge_kernel():
+  """Construct a set of 3x3 kernels for edge computation."""
+  filters = np.zeros((8, 3, 3, 1))
+  filters[:, 1, 1, :] = 1
+  offsets = [-1, 0, 1]
+  j = 0
+  for i, (dx, dy) in enumerate(itertools.product(offsets, repeat=2)):
+    if dx == 0 and dy == 0:
+      j += 1
+      continue
+    filters[i - j, 1 + dx, 1 + dy, 0] = -1
+  return tf.constant(filters, dtype=tf.float32)
+
+
+edge_filters = edge_kernel()
+
+
+def sub2ind(cols, r, c):
```

Comment:
are these always columns? Aren't you working on rows + columns at once? Maybe name this coord and give better names to 'r' and 'c' - these could be size and offset for instance, but I think you can find a better name for them.

Reply:
It converts from (x,y) indices to sequential indices in row-major order. I have changed r and 'c' to 'y' and 'x', and 'cols' is now named 'range'.